### ISSUE TITLE
Out-of-stock orders remain in "cart" state

### ISSUE THREAD

**Issue by filipefurtad0** [openfoodfoundation/openfoodnetwork#8177]
## Description

Orders which are not completed due to stock restrictions are left in a `cart` state - although no items are in the cart. Only after an admin finds the order and visits the respective order edit page will the state transit to `address`.

## Expected Behavior

Incomplete orders due to stock restrictions remain in the `Address` state.

## Actual Behaviour

Incomplete orders due to stock restrictions remain in the `Cart` state. After the admin clicks on the order on the UI the order transitions to the correct state (`Address`).

## Steps to Reproduce

As a customer
1. Add items to the cart and proceed to checkout

As an admin
2. Reduce the stock of that line item to zero

As a customer
3. Click "checkout" -> You should get redirected to the `/cart` page

As an admin
4. Find the order on the `/orders` page. Notice it is in the `cart` state, when displayed in the order list.
5. Click on that entry, i.e. visit `/admin/orders/<order_number>/edit` for the respective order
6. Notice it transitioned to the `Address` state

## Animated Gif/Screenshot

![Peek 2021-09-13 10-25](https://user-images.githubusercontent.com/49817236/133059684-4d2b59fe-7f33-465d-958a-8b52f3d5c4a0.gif)

## Workaround

Users will not be aware of this, but the workaround is to kook for a `cart` order which is has zero line items, when searching for incomplete orders. These are the only such orders appearing in the results list after PR #8148. 

## Severity

bug-s3: a feature is broken but there is a workaround

## Your Environment

* Version used: v4.1.1
* Browser name and version: Firefox 91
* Operating System and version (desktop or mobile): Ubuntu 20.04

**Comment by jaycmb** [comment]
> look for a cart order which is has zero line items, when searching for incomplete orders.

You mean orders that have 0 line items or orders that have line items with quantity=0?
I think the latter, right?

**Comment by filipefurtad0** [comment]
I was referring to orders with no line items, but I don't think we can have orders with line items with quantity zero - at least as displayed in the UI.

Interestingly, I think the issue is caused by line item which is out of stock (seen in the DB, quantity = 5) which does not appear in the UI for the admin. Only an added line item (with stock, quantity = 2) appears:
![image](https://user-images.githubusercontent.com/49817236/133247217-fda12c07-bc98-44a7-862e-f43375b3239a.png)

**Comment by jaycmb** [comment]
Yeah I meant whats described in #8178:) 

The problem also occurs to orders that have several items which are in stock, but item that is out of stock is blocking the rest of the order to be processed.
So the missing items remains quantity=0 on the user side but is not visible on the Admin Interface

**Comment by lin-d-hop** [comment]
I'm not convinced by this issue... why would an incomplete order with no line items go to the Address state? Why is that correct?

My understanding of the Address state is that it is only a possible when an order is populated in cart but does not have a user associated yet (which generally only happens in the BO).

What am I missing @filipefurtad0 ?

**Comment by filipefurtad0** [comment]
Yeah these issues are tricky as there is no user guide on the expected behavior. So a lot of what should be was discovered empirically and keeping in mind that the flows (backoffice and shopfront) order flows should be as similar as possible.

So there is plenty of room for interpretation here :sweat_smile: 

However:
> an order is populated in cart but does not have a user associated yet (which generally only happens in the BO)

Currently, this :point_up: generates orders in the `cart` state. If it should be `address` state then we'd need an issue for this. Below, an order created through the BO:

```

   id   |   number   | item_total | total | state | adjustment_total | user_id |         created_at         |         updated_at         | completed_at | bill_address_id | ship_address_id | payment_total | shipment_state | payment_state | email | special_instructions | distributor_id | order_cycle_id | currency | last_ip_address | customer_id | created_by_id | included_tax_total | additional_tax_total 
--------+------------+------------+-------+-------+------------------+---------+----------------------------+----------------------------+--------------+-----------------+-----------------+---------------+----------------+---------------+-------+----------------------+----------------+----------------+----------+-----------------+-------------+---------------+--------------------+----------------------
 453397 | R547585162 |       2.00 |  2.00 | cart  |             0.00 |         | 2022-02-09 15:53:04.835415 | 2022-02-09 15:53:17.243251 |              |                 |                 |          0.00 |                |               |       |                      |           1799 |           3516 | CAD      |                 |             |               |               0.00 |                 0.00
(1 row)
```

So to me (again, this being an empirical interpretation) an order in the address state is an order which _does_ have a customer and an address... but no items. This can't be attained through the BO orders (as far as I know).

But please feel free to close the issue, if you feel this is not the path to go :+1:

**Comment by lin-d-hop** [comment]
Hrm... I actually think we need to be looking at removing the 'Address' state completely. It doesn't seem to be useful but needs more investigation.

**Comment by Matt-Yorkley** [comment]
Quick notes:

- In relation to orders that fail due to items going out of stock which are then edited by an admin; we had some previous issues around that which were fixed recently by Cillian.
- Some of the complexity around these state changes is due to weirdness in the way the current checkout works. It does things like resetting the order back to previous states if it encounters an error, and to be honest it's a bit _messy_. The split checkout will work a bit differently.
